### PR TITLE
feat: integrate Milvus Vector Database

### PR DESCRIPTION
**Pull request by jinhonglin-ryan**
## Overview

This PR integrates Milvus as a Vector Database option into the Dynamiq platform, allowing for efficient storage and retrieval of high-dimensional vector embeddings. 

### Key Changes

1. **Milvus Connection and Configuration**
   - Created a new `MilvusConnection` class in `dynamiq/connections`, allowing configuration through deployment types (e.g., `LOCAL_FILE` for local development and `CLOUD` for remote).
   - Implemented `MilvusDeploymentType` to support various Milvus deployment options.

2. **Milvus Vector Store**
   - Added `MilvusVectorStore` class in `dynamiq/storages/vector` to support CRUD operations (create, retrieve, update, delete) on vectors stored in Milvus.

3. **Milvus Document Retriever and Writer Nodes**
   - Added `MilvusDocumentRetriever` node to perform vector similarity searches with optional filters and a specified top-k retrieval limit.
   - Added `MilvusDocumentWriter` node to allow for seamless document ingestion and vector indexing.

4. **Unit and Integration Tests**
   - Wrote unit tests for the Milvus vector store, retriever, and writer nodes to ensure reliability and consistency.
   - Implemented integration tests to validate end-to-end workflows with Milvus as the vector database backend.

### CODE REVIEW

**Comment by jinhonglin-ryan** [comment]
@delamainer 
Hi, could you please review this pr and see if there are any further adjustments needed? Thanks!

**Comment by delamainer** [comment]
> @delamainer Hi, could you please review this pr and see if there are any further adjustments needed? Thanks!

@jinhonglin-ryan thank you for your contribution! @acoola @TrachukT will help you here.

**Comment by jinhonglin-ryan** [comment]
Hi @acoola and @TrachukT,

It looks like the test failure may stem from a missing Pinecone API key, which seems unrelated to the Milvus integration I've implemented. Could we verify that the Pinecone API settings are correctly configured? Please let me know if there’s anything else you’d like me to adjust on my end.

Thank you!

**Comment by TrachukT** [comment]
> Hi @acoola and @TrachukT,
> 
> It looks like the test failure may stem from a missing Pinecone API key, which seems unrelated to the Milvus integration I've implemented. Could we verify that the Pinecone API settings are correctly configured? Please let me know if there’s anything else you’d like me to adjust on my end.
> 
> Thank you!

Hi @jinhonglin-ryan, that`s the problem from our side connected to tests with credentials, as far as I know @acoola already working on it

File: dynamiq/connections/connections.py [hunk 1]
```diff
@@ -855,6 +856,61 @@ def connect(self):
...
+    """
+    Defines various deployment types for different Milvus deployments.
+
+    Attributes:
+        LOCAL_FILE (str): Represents a local file-based deployment.
+        DOCKER (str): Represents a Docker-based deployment on localhost.
+        ZILLIZ_CLOUD (str): Represents a cloud deployment on Zilliz Cloud.
+    """
+
+    LOCAL_FILE = "local_file"
```
Comment (acoola) [inline dynamiq/connections/connections.py:869]: Those types are so specific and don't cover other possible deployments variations.
So let's use more general like `FILE` and `HOST`:
- `FILE` - local file with validation for `.db` suffix
- `HOST` could be any cloud, cluster, or single machine deployment with or without token. So all validation could be done by `MilvusClient`. In `HOST` case client connect would looks like:
 ```python
 milvus_client = MilvusClient(uri=self.uri, token=self.api_key)
 ```

File: dynamiq/connections/connections.py [hunk 2]
```diff
@@ -855,6 +856,61 @@ def connect(self):
...
+
+    type: Literal[ConnectionType.Milvus] = ConnectionType.Milvus
+    deployment_type: MilvusDeploymentType = MilvusDeploymentType.LOCAL_FILE
+    uri: str = Field(default_factory=partial(get_env_var, "MILVUS_URI", "http://localhost:19530"))
+    api_key: str | None = Field(default_factory=partial(get_env_var, "MILVUS_API_TOKEN", None))
+
+    def connect(self):
+        from pymilvus import MilvusClient
+
+        if self.deployment_type == MilvusDeploymentType.LOCAL_FILE:
```
Comment (acoola) [inline dynamiq/connections/connections.py:893]: Let's move `FILE` validation to Pydantic model validator to do it on initialization step. Could be like:
```python
@model_validator(mode="after")
def validate_uri(self) -> "Milvus":
    if self.self.deployment_type == MilvusDeploymentType.LOCAL and not self.uri.endswith(".db"):
        raise ValueError("For local file deployment, URI should point to a local file ending with '.db'.")
    return self
 ```

**Comment by acoola** [comment]
@jinhonglin-ryan Thank you for your great work and contribution. 
Left a few minor comments. Please also rebase `main` to make it up to date

**Comment by jinhonglin-ryan** [comment]
Hi @acoola,

Thanks for your feedback. I have adjusted the deployment type based on your suggestions, and also made sure the unit testings are aligned with the changes. Also I have changed to using a tuple when inserting metadata. Please feel free to review these changes and let me know if there are further possible improvements! Much appreciated for your time and feedback. Hope to hear back from you soon!

File: tests/integration/nodes/retrievers/test_milvus_retriever_flow.py [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+@pytest.fixture
+def mock_milvus_client():
+    return MagicMock()
+
+
+@pytest.fixture
+def mock_milvus_connection():
+    return MilvusConnection(
+        deployment_type=MilvusDeploymentType.LOCAL_FILE,
```
Comment (acoola) [inline tests/integration/nodes/retrievers/test_milvus_retriever_flow.py:23]:
```suggestion
        deployment_type=MilvusDeploymentType.FILE,
```

File: tests/integration/nodes/writers/test_milvus_writer_flow.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+def mock_milvus_client():
+    mock_client = MagicMock()
+    mock_client.has_collection.return_value = True
+    return mock_client
+
+
+@pytest.fixture
+def mock_milvus_connection():
+    return MilvusConnection(
+        deployment_type=MilvusDeploymentType.LOCAL_FILE,
```
Comment (acoola) [inline tests/integration/nodes/writers/test_milvus_writer_flow.py:25]:
```suggestion
        deployment_type=MilvusDeploymentType.FILE,
```

**Comment by jinhonglin-ryan** [comment]
Hi @acoola,

Thank you for the feedback! I have adjusted testing based on the new deployment type definition. Please let me know if there are further adjustments needed! Thank you!

**Review by acoola — APPROVED** [review]
Thank you for your great work and contribution

**Pull request merged**